vent/controller/control_module.py

```diff
@@ -98,144 +115,165 @@ def test_critical_levels(self, min, max, value, name):
 
             # get the pressure niveau heuristically (much faster than fitting)
             # 20 and 80 percentiles pulled out of my hat.
-            self.DATA_PEEP = np.percentile(pressure, 20)
-            self.DATA_PIP = np.percentile(pressure, 80)
+            self._DATA_PEEP = np.percentile(pressure, 20)
+            self._DATA_PIP = np.percentile(pressure, 80)
 
             # measure time of reaching PIP, and leaving PIP
-            self.DATA_PIP_TIME = phase[np.min(np.where(pressure > self.DATA_PIP))]
-            self.DATA_I_PHASE = phase[np.max(np.where(pressure > self.DATA_PIP))]
+            self._DATA_PIP_TIME = phase[np.min(np.where(pressure > self._DATA_PIP))]
+            self._DATA_I_PHASE = phase[np.max(np.where(pressure > self._DATA_PIP))]
 
             # and measure the same for PEEP
-            self.DATA_FIRST_PEEP = phase[np.min(np.where(np.logical_and(pressure < self.DATA_PEEP, phase > 1)))]
-            self.DATA_BPM = 60. / phase[-1]  # 60 sec divided by the duration of last waveform
+            self.__DATA_FIRST_PEEP = phase[np.min(np.where(np.logical_and(pressure < self._DATA_PEEP, phase > 1)))]
+            self._DATA_BPM = 60. / phase[-1]  # 60 sec divided by the duration of last waveform
 
-            self.test_critical_levels(min=self.PIP_min, max=self.PIP_max, value=self.DATA_PIP, name="PIP")
-            self.test_critical_levels(min=self.PIP_time_min, max=self.PIP_time_max, value=self.DATA_PIP_TIME, name="PIP_TIME")
-            self.test_critical_levels(min=self.PEEP_min, max=self.PEEP_max, value=self.DATA_PEEP, name="PEEP")
-            self.test_critical_levels(min=self.bpm_min, max=self.bpm_max, value=self.DATA_BPM, name="BREATHS_PER_MINUTE")
-            self.test_critical_levels(min=self.I_phase_min, max=self.I_phase_max, value=self.DATA_I_PHASE, name="I_PHASE")
+            self.__test_critical_levels(min=self.__PIP_min, max=self.__PIP_max, value=self._DATA_PIP, name="PIP")
+            self.__test_critical_levels(min=self.__PIP_time_min, max=self.__PIP_time_max, value=self._DATA_PIP_TIME, name="PIP_TIME")
+            self.__test_critical_levels(min=self.__PEEP_min, max=self.__PEEP_max, value=self._DATA_PEEP, name="PEEP")
+            self.__test_critical_levels(min=self.__bpm_min, max=self.__bpm_max, value=self._DATA_BPM, name="BREATHS_PER_MINUTE")
+            self.__test_critical_levels(min=self.__I_phase_min, max=self.__I_phase_max, value=self._DATA_I_PHASE, name="I_PHASE")
 
     def get_sensors(self) -> SensorValues:
         # This will depend on simulation vs. reality
         pass
 
     def get_alarms(self) -> List[Alarm]:
         # Returns all alarms as a list
-        ls = self.logged_alarms
-        for alarm_key in self.active_alarms.keys():
-            ls.append(self.active_alarms[alarm_key])
-        return ls
+        lock.acquire()
+        new_alarm_list = self.__logged_alarms.copy()     # Make sure to return a copy!
```

Comment:
I don't think we want to be accessing these "__" underscore variables in the get_ set_ functions. The way to think about this is consider there are two threads, one is the controller thread, and another thread is a customer of the controller (for example a UI thread). The UI thread is calling the get_ set_ functions and that thread is running the code within those functions (even though the code belongs to the controller module it is being run by the UI thread). So we need to coordinate access to variables in these functions that both threads are accessing so they don't clobber each other. If we access the __ underscore variables in the get_ set_ functions then we need to have locks on those variables everywhere in the controller they are accessed. So I think it might be simpler to have a copy of the __ variables that is only used in get_ set_ functions and updated periodically by the controller, i.e. the shared_ variables.

Reply:
I'm on it, no worries. Just a lot of typing... :-)